plugins/catalog-backend-module-ldap/src/ldap/org.ts

```diff
@@ -14,25 +14,29 @@
  * limitations under the License.
  */
 
-import { GroupEntity, UserEntity } from '@backstage/catalog-model';
+import {
+  GroupEntity,
+  stringifyEntityRef,
+  UserEntity,
+} from '@backstage/catalog-model';
 
 // TODO: Copied from plugin-catalog-backend, but we could also export them from
 // there. Or move them to catalog-model.
 
 export function buildOrgHierarchy(groups: GroupEntity[]) {
-  const groupsByName = new Map(groups.map(g => [g.metadata.name, g]));
+  const groupsByRef = new Map(groups.map(g => [stringifyEntityRef(g), g]));
 
   //
   // Make sure that g.parent.children contain g
   //
 
   for (const group of groups) {
-    const selfName = group.metadata.name;
-    const parentName = group.spec.parent;
-    if (parentName) {
-      const parent = groupsByName.get(parentName);
-      if (parent && !parent.spec.children.includes(selfName)) {
-        parent.spec.children.push(selfName);
+    const selfRef = stringifyEntityRef(group);
+    const parentRef = group.spec.parent;
+    if (parentRef) {
+      const parent = groupsByRef.get(parentRef);
+      if (parent && !parent.spec.children.includes(selfRef)) {
```

Comment:
So! This is all looking nice but I realized something else.

There's a lot of `Array.includes` in here right? Those are actually not correct now. Or at least, they may lead to duplicates.

The `spec` fields are forgiving. They allow the user to skip kind and namespace, mostly, if they are of the default kind and in the same namespace as the entity itself. So the code that reads these and translates them into entity relations give some defaults to `parseEntityRef`. This means that if e.g. `parent.spec.children` contained `foo`, then you'd introduce the duplicate `group:default/foo` here. Not a huge deal - nothing will break - but it'll look weird for the user since they wrote a spec by hand and then there are odd duplicates in the final result.

I think the right thing to do here is to do

```ts
!parent.spec.children.some(
  c => selfRef === parseEntityRef(c, {
    defaultKind: 'Group',
    defaultNamespace: group.metadata.namespace || DEFAULT_NAMESPACE
  }))
```

the list of defaults etc is in this authoritative doc https://backstage.io/docs/features/software-catalog/descriptor-format

Would you be up for updating these occurrences of `.includes`? Maybe with a backing test where the spec DID have a shorthand field in there from the start and it doesn't get replaced or get a dupe

Reply:
Thanks for the feedback and additional details @freben. I may be missing something, but from what I can tell the catalog-backend-module-ldap plugin builds out these groups/user entities from scratch each time it loads in data. I don't see where it would read in previously existing Users/Groups and then append new information based on the LDAP data that it receives, so I'm unsure how to recreate the behavior you're describing in test cases.

Based on my understanding, an entity loaded from the LDAP dataset that used to have a `parent.spec.children` that contained `foo`  would be built out with a `parent.spec.children` value of `group:default/foo` instead with the new logic. 

Reading in the LDAP groups:
https://github.com/backstage/backstage/blob/0df260a54fbb8c193dbedda7379a60a5dab0ff85/plugins/catalog-backend-module-ldap/src/ldap/read.ts#L204-L250

Creating new entities for each Group entry, no reference to existing Users/Groups from what I can see:
https://github.com/backstage/backstage/blob/0df260a54fbb8c193dbedda7379a60a5dab0ff85/plugins/catalog-backend-module-ldap/src/ldap/read.ts#L140-L158

New relationship creation based on `entityRef` rather than `name` fields:
https://github.com/ericfitzpatrick/backstage/blob/b37bca4e4e2d6103f670f6d23184fe47fbccaccd/plugins/catalog-backend-module-ldap/src/ldap/read.ts#L398-L473

I tried to define in a YAML file one of the groups/users that exists in the test LDAP dataset that I'm using, and Backstage detects it as a conflict and the LDAP data took priority (probably because it loaded first): 

```
[1] 2022-04-12T15:26:13.653Z catalog info Read 16 LDAP users and 3 LDAP groups in 1.5 seconds. Committing... type=plugin target=ldap://ldap.forumsys.com class=LdapOrgEntityProvider taskId=LdapOrgEntityProvider:our-ldap-master:refresh taskInstanceId=ae481a93-cce2-41fb-8639-1251e12b5666
[1] 2022-04-12T15:26:13.668Z catalog info Committed 16 LDAP users and 3 LDAP groups in 0.0 seconds. type=plugin target=ldap://ldap.forumsys.com class=LdapOrgEntityProvider taskId=LdapOrgEntityProvider:our-ldap-master:refresh taskInstanceId=ae481a93-cce2-41fb-8639-1251e12b5666
[1] 2022-04-12T15:26:13.694Z catalog warn Detected conflicting entityRef group:default/scientists already referenced by ldap-org-provider:our-ldap-master and now also file:<redacted>\packages\catalog-model\examples\acme\scientists.yaml type=plugin
[1] 2022-04-12T15:26:13.694Z catalog warn Detected conflicting entityRef user:default/einstein already referenced by ldap-org-provider:our-ldap-master and now also file:<redacted>\packages\catalog-model\examples\acme\scientists.yaml type=plugin
```

Please let me know if there's a way for the LDAP data and user-defined data to coexist that I'm missing. If so, I can make the changes you suggested and build a new test case once I can recreate the behavior you described locally.

Reply:
Right, but the code we're looking at here only amends existing data, when necessary. The "existing" data is stuff that's copied verbatim from LDAP entries' attributes. So if an LDAP entry has children, then the generated entity will contain some `spec.children`. Then later on, when this wire-up code tries to ensure consistency, it'll add a long-form full duplicate. Right?

Reply:
I think my confusion here is that we create those fields in resolveRelations based on the LDAP data, rather than the LDAP data being set to `spec.children`, `spec.parent`, etc. verbatim. The stuff that is copied verbatim in the group transformer are things like name, description, etc., but for group <-> group relationships (`spec.children`, `spec.parent`) and user <-> group relationships (`spec.memberOf`) that's all built out in resolveRelations which (as part of the changes I made) will always make those fields in the form of entityRef to my understanding.

resolveRelations setting these fields after they're generated with the entityRef format earlier in the function:
https://github.com/ericfitzpatrick/backstage/blob/ebc0aef2c6203cb7a2edfc6ac7adaf109fa88af9/plugins/catalog-backend-module-ldap/src/ldap/read.ts#L450-L470

The below tests are passing with the current .includes() checks which simulates an LDAP group having a child group:
https://github.com/ericfitzpatrick/backstage/blob/ebc0aef2c6203cb7a2edfc6ac7adaf109fa88af9/plugins/catalog-backend-module-ldap/src/ldap/read.test.ts#L346-L389

Reply:
ohh man you are totally right, I should read more carefully before commenting - sure, thanks for sticking with me!